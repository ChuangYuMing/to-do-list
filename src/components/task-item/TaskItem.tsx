import React, { useState, useId } from "react";
import { Task } from "@/types";
import styles from "./TaskItem.module.scss";

type TaskItemProps = {
  task: Task;
  updateTask: (taskId: number, payload: Partial<Task>) => void;
  deleteTask: (taskId: number) => void;
};

export default function TaskItem({
  task,
  updateTask,
  deleteTask,
}: TaskItemProps) {
  const id = useId();
  const [isEditing, setIsEditing] = useState<boolean>(false);

  const formatDate = (timestamp: number) => {
    const date = new Date(timestamp);
    const formattedDate = date.toLocaleDateString("en-US", {
      year: "numeric",
      month: "short",
      day: "numeric",
    });
    const formattedTime = date.toLocaleTimeString("en-US", {
      hour: "2-digit",
      minute: "2-digit",
      second: "2-digit",
      hour12: false,
    });
    return `${formattedDate}, ${formattedTime}`;
  };

  return (
    <div key={task.createdAt} className={styles["task-item"]}>
      <div className={styles["task-content"]}>
        <input
          id={id}
          type="checkbox"
          checked={task.completed}
          onChange={(e) => updateTask(task.id, { completed: e.target.checked })}
          className={styles["task-checkbox"]}
        />
        <label htmlFor={id} className={styles["checkbox-label"]}></label>
        {isEditing ? (
          <input
            type="text"
            value={task.description}
            onChange={(e) =>
              updateTask(task.id, { description: e.target.value })
            }
            onBlur={() => setIsEditing(false)}
            className={styles["task-description-input"]}
          />
        ) : (
          <span className={styles["task-description"]}>{task.description}</span>
        )}
      </div>
      <div className={styles["task-actions"]}>
        <span
          className={styles["task-action"]}
          onClick={() => setIsEditing(true)}
        >
          Edit
        </span>
        <span
          className={styles["task-action"]}
          onClick={() => deleteTask(task.id)}
        >
          Delete
        </span>
      </div>
      <span className={styles["created-date"]}>
        {formatDate(task.createdAt)}
      </span>
    </div>
  );
}
