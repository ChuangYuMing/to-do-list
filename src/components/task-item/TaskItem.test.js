import React from "react";
import { render } from "@testing-library/react";
import TaskItem from "./TaskItem";

it("renders correctly", () => {
  const mockTask = {
    createdAt: 1,
    id: 1,
    description: "Task 1",
    completed: false,
  };

  const { asFragment } = render(
    <TaskItem
      key={mockTask.createdAt}
      task={mockTask}
      updateTask={() => {}}
      deleteTask={() => {}}
    />
  );
  expect(asFragment()).toMatchSnapshot();
});
