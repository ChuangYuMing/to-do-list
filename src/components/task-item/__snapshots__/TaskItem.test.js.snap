// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<DocumentFragment>
  <div
    class="task-item"
  >
    <div
      class="task-content"
    >
      <input
        class="task-checkbox"
        id=":r0:"
        type="checkbox"
      />
      <label
        class="checkbox-label"
        for=":r0:"
      />
      <span
        class="task-description"
      >
        Task 1
      </span>
    </div>
    <div
      class="task-actions"
    >
      <span
        class="task-action"
      >
        Delete
      </span>
    </div>
    <span
      class="created-date"
    >
      Jan 1, 1970, 08:00:00
    </span>
  </div>
</DocumentFragment>
`;
